.gitbook/assets/ReadonlyREST-api.yml

```diff
@@ -307,6 +386,46 @@ components:
         status: FILE_SIZE_EXCEEDS
         message: "The file size exceeds the maximum 10MB allowed limit. Please upload a smaller file."
 
+    UpdateLicenseKeySuccessResponse:
+      summary: License key updated with success
+      value:
+        status: SUCCESS
+        message: "ok"
+    UpdateLicenseKeyUnmodifiedResponse:
+      summary: License key not modified
+      value:
+        status: UNMODIFIED
+        message: "License token already exists is not modified. If you want to overwrite your license key use overwrite=true query string parameter."
+    UpdateLicenseKeyExpiredResponse:
+      summary: License key expired
+      value:
+        status: JWT_EXPIRED
+        message: "The JWT token provided has expired."
+    UpdateLicenseKeyMalformedResponse:
+      summary: License key malformed
+      value:
+        status: JWT_MALFORMED
```

Comment:
Continuing the thread ... if the answer is "yes, JWT is the implementation detail", should we explain in the statuses what's wrong with the JWT? We can write in the error log what is exactly wrong, but in API it's better to be generic, if user cannot (or should not) understand what we mean.

Reply:
> if the answer is "yes, JWT is the implementation detail", should we explain in the statuses what's wrong with the JWT?

I don't understand this sentence. Do you mean "If we established that it's okay to use JWT token information in our external API, let's not try to describe in the status what's going on"?


> should we explain in the statuses what's wrong with the JWT?

Okay, so the proposition is to use BAD_REQUEST for all 400 errors and describe in the message, what's the exact problem (what we already did). 

> but in API it's better to be generic

I don't have a lot of experience with exposing public API, so my question is maybe a little stupid, but why it's better to be generic? I assume that if someone, wants to use our external API in the "automated way", he/she will want to react to the specific situation automatically, based on the status. So In our situation, if the `JWT_EXPIRED` token, status is, let's send an email (or something like this. :P)

> if user cannot (or should not) understand what we mean.

The question is, is it true in our case?  I mean, IMO since it's a technical user he/she will know what's going on here, and I don't see any problems with exposing this kind of information to the end user, day you have on mind specific case?

Reply:
> I don't understand this sentence. Do you mean "If we established that it's okay to use JWT token information in our external API, let's not try to describe in the status what's going on"?

I mean that using status `JWT_MALFORMED` we say sth like this: "hey, the JWT you uploaded is malformed", but IMO we should say: "hey, the license you uploaded is malformed". So, IMO `MALFORMED_LICENSE` would be better.

> Okay, so the proposition is to use BAD_REQUEST for all 400 errors and describe in the message, what's the exact problem (what we already did).

Yes, using the message is better, because the message text cannot be treated as a part of the API. The message can easily be changed in the future. But the message should follow the same rules the API should follow: we should not mention about things that users are not aware of or about sth they are able to understand (after research) but what doesn't bring anything useful to the context.

> I don't have a lot of experience with exposing public API, so my question is maybe a little stupid, but why it's better to be generic? I assume that if someone, wants to use our external API in the "automated way", he/she will want to react to the specific situation automatically, based on the status. So In our situation, if the JWT_EXPIRED token, status is, let's send an email (or something like this. :P)

No, the question is a good one. It's all about maintenance and backward compatibility of the API. Let's assume that in the future we decide to hide information passed in the JWT token and we decide to encrypt it. It's not a JWT anymore. But it's still a license. But in the API we have JWT all over the place :) What we can do? We can break the compatibility and remove the "JWT" abbrev from the API or not break it and decide to live with sth which is not true anymore. 

During the designing of the API, we were focused too much on the current implementation, but always we should focus on the business side of the API and take into consideration the knowledge the user has (or needs to have).

> The question is, is it true in our case? I mean, IMO since it's a technical user he/she will know what's going on here, and I don't see any problems with exposing this kind of information to the end user, day you have on mind specific case?

Let's determine how users work with the ROR license (correct me when I'm wrong). 
1. they go to the Customer Portal and generate a license
2. they copy the license (the long string)
3. they upload the license
  a) using the ROR settings page
  b) using API (the one decided here)
 
So the facts are:
* we generate the license 
* user uploads it
* user doesn't have to know what implementation for the license structure

What could go wrong?
1. user copies part of the license (the common issue) - we should tell him: "we detect that sth is wrong with the license". There is no sense in telling them what, because they won't be able to fix it. At least we can advice him to check if he copied the whole string
2. user uses an expired license (a common issue too) - we should tell him: "hey, you used the expired license. But the new one for a god sake! :)". We don't need to tell them, that the JWT is expired. Their ROR license is expired. They context is an ROR license, not a JWT. 
3. user tries to modify the license (not so common thing) - we should tell him the same we do in the case of point 1. 

As you can see there is no need to bother them about the fact that JWT signature is bad, or JWT is invalid. 
This is why I think we should remove all the JWT-related things from this API.

Reply:
Okay, so your proposition, is to remove states that are let's say, programmatic errors, because for some reason we generated a bad license key, and don't expose it to the end user, but still log it as a console error.  I misunderstood you here, I thought, that your proposed was about removing all 400 activation key based statuses. Changed ✔️

Reply:
yes, it's still worth logging it for the support purposes 👍